### PR TITLE
[#162413][Daily Rate] Monthly calendar UI for daily booking instruments

### PR DESCRIPTION
**Pull request by joaquinco**
# Notes

Render monthly view for daily booking instrument's reservation page

## Additional Context

- Create class `CalendarEventsPresenter` in charge of rendering calendar events for reservations and schedule rules
- Move `ScheduleRuleCalendarPresenter` to `ScheduleRules::CalendarPresenter` and add the possibility to handle month calendar view, which for now differs on how unavailable days are rendered as a full gray box.
- Allow full calendar init to receive options through html element data attributes
- Only show monthly view for daily booking instruments' schedule (Product Schedule admin tab)
- Use `as_json` instead of `to_json` for calendar event presenters which better follows rails semantics (return Hash or Array of hashes vs json formatted string)

## Testing Notes

Go to the reservation page of a daily booking instrument.
The calendar is not editable.
When changing the month on the start date the calendar renders the selected month events.

## Screenshot

Captura desde 2024-11-07 16-28-58

### CODE REVIEW

File: app/assets/javascripts/app/calendar.js [hunk 1]
```diff
@@ -49,6 +54,26 @@ window.FullCalendarConfig = class FullCalendarConfig {
...
+    let self = this;
+    allowedKeys.forEach(function(key) {
+      let value = self.$element.data(key)
+      if (value) {
+        ret[key] = value;
+      }
+    });
+
+    console.log(this.$element.data());
+    console.log(ret);
```
Comment (LeticiaErrandonea) [inline app/assets/javascripts/app/calendar.js:73]: nit: stray `console.log`s

File: app/assets/javascripts/instruments.js [hunk 2]
```diff
@@ -1,9 +1,15 @@
 $(document).ready(function() {
-  new FullCalendarConfig($("#calendar"), {
-    header: {
-      left: 'title',
-      center: '',
-      right: 'prev,next today agendaDay,agendaWeek,month',
-    }
-  }).init();
+  const calendar = $("#calendar");
+  let defaultView = calendar.data('defaultView');
```
Comment (LeticiaErrandonea) [inline app/assets/javascripts/instruments.js:3]: nit: I think this can be a `const`

File: app/assets/javascripts/instruments.js [hunk 3]
```diff
@@ -1,9 +1,15 @@
...
+
+  const header = { left: 'title', center: '', right: '' };
+
+  if (defaultView != 'month') {
+    header.right = 'prev,next today agendaDay,agendaWeek,month';
+  }
+
+  new FullCalendarConfig(
+    calendar,
+    { header: header }
```
Comment (LeticiaErrandonea) [inline app/assets/javascripts/instruments.js:13]: nit: 
```suggestion
    { header }
```

File: app/controllers/reservations_controller.rb [hunk 4]
```diff
@@ -16,6 +16,8 @@ class ReservationsController < ApplicationController
   include ReservationSwitch
   helper TimelineHelper
 
+  MAX_RESERVATIONS_PERIOD = 40.days
```
Comment (LeticiaErrandonea) [inline app/controllers/reservations_controller.rb:19]: Was this number already used somewhere?

File: app/helpers/calendar_helper.rb [hunk 5]
```diff
@@ -0,0 +1,26 @@
+# frozen_string_literal: true
+
+module CalendarHelper
+
+  def product_calendar_view(product)
+    return "month" if product.daily_booking?
+
+    "agendaWeek"
```
Comment (LeticiaErrandonea) [inline app/helpers/calendar_helper.rb:8]: nit: in this cases I usually go for a regular if-else or even a ternary. When it's just to return a value, I find any of them more readable than this pattern

**Review by LeticiaErrandonea — COMMENTED** [review]
Left a few comments. Specs are failing so I didn't go in depth

File: app/controllers/reservations_controller.rb (hunk 4 above)
Comment (joaquinco) [reply app/controllers/reservations_controller.rb:19]: Nope, but if it the index endpoint is only used for calendars it should be enough

**Comment by LeticiaErrandonea** [comment]
I did some Dev QA and I noticed something that I think is part of the Schedule Rule ticket, but just in case: it'd be better to have this view display the monthly calendar. Bringing this up because PR updates calendars
image

Other than that, it seemed to work fine, for both Daily booking and existing instruments

**Comment by joaquinco** [comment]
Thanks for the QA.

I thought about changing the schedule rules calendar view but since they are defined for each day of the week I though they were easier to visualize the way it was already done.

**Review by LeticiaErrandonea — APPROVED** [review]
LGTM!

**Pull request merged**